# Ödev: Bir E-Ticaret Şirketi İçin Veri Analizi ve Raporlama


**Pair-11**

* Ayşenur Yörür aysenuryorur0@gmail.com
* Buse Erarslan erarslanbuse01@gmail.com
* Şerife Kaya kayaseriife@gmail.com
* Sena Sücüllü senasucullu@gmail.com

## Ödev Açıklaması:

Bir e-ticaret şirketinde çalıştığınızı hayal edin. Şirket, müşterilerin alışveriş alışkanlıklarını anlamak, en çok satan ürünleri belirlemek ve kârlılığı artırmak için veri analizi yapmanızı istiyor. Size verilen satış verilerini NumPy kullanarak analiz etmeniz gerekiyor.

1. Veri Seti (Üretilmesi Gereken Rastgele Veriler)
Her öğrenci rastgele üretilmiş verilere dayalı analiz yapmalıdır. Aşağıdaki verileri NumPy kullanarak oluşturun:

* 1000 müşterinin harcama tutarları (50 TL ile 5000 TL arasında rastgele)
* 1000 müşterinin satın aldığı ürün sayıları (1 ile 10 arasında rastgele)
* 1000 müşterinin indirimden yararlanıp yararlanmadığı (0: Hayır, 1: Evet - %30 * indirim aldı)
* Ürün fiyatları (10 TL ile 1000 TL arasında rastgele)
* Kategori bazlı satış adetleri (5 kategori için rastgele satış sayıları)

2. İstenilen Analizler
Aşağıdaki analizleri gerçekleştirin:

  A. Temel İstatistikler
    * Ortalama müşteri harcaması nedir? (mean)
    * En yüksek ve en düşük harcama miktarları nedir? (max, min)
    * Kaç müşteri indirim kullanmıştır? (sum)

  B. Satış Analizi
    * İndirim kullanan müşterilerin ortalama harcaması ile indirim kullanmayanların ortalama harcaması arasındaki farkı bulun.
    * Toplam satış gelirini hesaplayın.
    * En çok satılan ürünleri ve kategorileri belirleyin.
    * Satın alınan ürün sayısı ile harcama arasında korelasyon var mı? (np.corrcoef() kullanarak)

  C. Sezonluk Karşılaştırma (Geçen Yılla Kıyaslama)
    * Geçen yılın verilerini rastgele oluşturarak bu yıl ile karşılaştırın. Satışlarda ortalama % kaç artış veya azalma olmuş?
    * Geçen yıl en çok satılan ürün ile bu yıl en çok satılan ürünü karşılaştırın.

  D. Stratejik Kararlar
    * Elde edilen verilere dayanarak e-ticaret şirketine 3 öneri sunun. (Örneğin, indirimlerin satışlara etkisini değerlendirerek)

## Veri Seti Üretimi

In [ ]:
import numpy as np
n = 1000

### **Müşterinin Harcama Tutarları**

1000 müşterinin harcama tutarları (50 TL ile 5000 TL arasında rastgele)

Seed consistency sağlıyor. Rastgele aynı değerlerin üretilmesinde rol alıyor.
Müşterinin harcama tutarları.

Örneğin
* 1. müşteri 910 TL
* 2. müşteri 3822 TL
* 3. müşteri 3142 TL

harcama yapmıştır

In [ ]:
np.random.seed(42)
total_spend = np.random.randint(50, 5000, n) #müşterinin harcama tutarları
print("İlk 5:",total_spend[0:5],"...")

İlk 5: [ 910 3822 3142  516 4476] ...


### **Müşterinin Satın Aldığı Ürün Sayıları**

Örneğin
* 1. müşteri 7 ürün almıştır.
* 2. müşteri 4 ürün almıştır.
* 3. müşteri 8 ürün almıştır.

Müşterilerin hangi ürünleri aldıklarını **bilmiyoruz**!

In [ ]:
np.random.seed(42)
num_products = np.random.randint(1, 10, n)  #müşterinin satın aldığı ürün sayıları
print("İlk 5:",num_products[0:5],"...") #İlk 5'i yazdır

İlk 5: [7 4 8 5 7] ...


### **Müşterinin İndirimden Yararlanıp Yararlanmadığı**

1000 müşterinin indirimden yararlanıp yararlanmadığı (0: Hayır, 1: Evet - %30 * indirim aldı)

Örneğin
* 1. müşteri 0 indrimden yararlanmamıştır.
* 2. müşteri 1 indrimden yararlanmıştır.
* 3. müşteri 1 indrimden yararlanmıştır.


In [ ]:
np.random.seed(42)
has_discount = np.random.choice([0, 1], n, p=[0.7, 0.3]) #müşterinin indirimden yararlanıp yararlanmadığı
print("İlk 5:",has_discount[0:5],"...") #İlk 5'i yazdır

İlk 5: [0 1 1 0 0] ...


### Ürün Fiyatları

Ürün fiyatları (10 TL ile 1000 TL arasında rastgele)

20 ürünümüz olsun ve fiyatları da aşağıdaki gibi olsun. Örneğin...
1. ürünün fiyatı 935,
2. ürünün fiyatı 930
3. ürünün fiyatı 368


In [ ]:
np.random.seed(42)
product_prices = np.random.randint(10, 1000, 20) #Ürün fiyatları
print(product_prices,"\n")

product_names = np.array([f"Product_{i+1}" for i in range(len(product_prices))])

# Ürün isimleri ve fiyatlarını birlikte gösterelim
for name, price in zip(product_names, product_prices):
    print(f"{name}: {price} TL")

[112 445 870 280 116  81 710  30 624 131 476 224 340 468  97 382 109 881
 673 140] 

Product_1: 112 TL
Product_2: 445 TL
Product_3: 870 TL
Product_4: 280 TL
Product_5: 116 TL
Product_6: 81 TL
Product_7: 710 TL
Product_8: 30 TL
Product_9: 624 TL
Product_10: 131 TL
Product_11: 476 TL
Product_12: 224 TL
Product_13: 340 TL
Product_14: 468 TL
Product_15: 97 TL
Product_16: 382 TL
Product_17: 109 TL
Product_18: 881 TL
Product_19: 673 TL
Product_20: 140 TL


### Product Sales, Her ürün kaçar adet satılmış?

In [ ]:
np.random.seed(42)
total_products = np.sum(num_products)

# Satış adetlerini rastgele belirle, toplam num_products olacak şekilde
product_sales = np.random.multinomial(total_products, np.random.dirichlet(np.ones(20)))

# Ürün isimlerini oluştur
product_names = np.array([f"Product_{i+1}" for i in range(20)])

# Sonuçları yazdıralım
for name, sales in zip(product_names, product_sales):
    print(f"{name}: {sales} adet satıldı")

Product_1: 111 adet satıldı
Product_2: 813 adet satıldı
Product_3: 350 adet satıldı
Product_4: 240 adet satıldı
Product_5: 43 adet satıldı
Product_6: 44 adet satıldı
Product_7: 10 adet satıldı
Product_8: 502 adet satıldı
Product_9: 257 adet satıldı
Product_10: 320 adet satıldı
Product_11: 5 adet satıldı
Product_12: 955 adet satıldı
Product_13: 470 adet satıldı
Product_14: 67 adet satıldı
Product_15: 61 adet satıldı
Product_16: 53 adet satıldı
Product_17: 125 adet satıldı
Product_18: 190 adet satıldı
Product_19: 176 adet satıldı
Product_20: 106 adet satıldı


### Kategoriler


In [ ]:
np.random.seed(42)

# Kategori isimlerini belirle
categories = np.array(["C1", "C2", "C3", "C4","C5"])

# Toplam ürün sayısına uygun olacak şekilde ürün kategorilerini belirle
product_categories = np.random.choice(categories, total_products)

In [ ]:
# Görüldüğü üzere toplam ürün sayısı ile kategoriler eşleştiler. Her ürünün kategorisi var artık.
total_products,product_categories.shape

(4898, (4898,))

In [ ]:
# Kategori bazlı satış adetlerini hesapla
unique_categories, category_sales_counts = np.unique(product_categories, return_counts=True)
category_sales = dict(zip(unique_categories, category_sales_counts))

# Sonuçları yazdıralım (ilk 20 ürünü gösterelim)
print(f"Product Name \t Category Name  Price    Sales Count")
for name, price, category, sales in zip(product_names, product_prices, product_categories, product_sales):
    print(f"{name} \t {category}  \t\t{price} TL \t {sales} adet satıldı")

Product Name 	 Category Name  Price    Sales Count
Product_1 	 C4  		112 TL 	 111 adet satıldı
Product_2 	 C5  		445 TL 	 813 adet satıldı
Product_3 	 C3  		870 TL 	 350 adet satıldı
Product_4 	 C5  		280 TL 	 240 adet satıldı
Product_5 	 C5  		116 TL 	 43 adet satıldı
Product_6 	 C2  		81 TL 	 44 adet satıldı
Product_7 	 C3  		710 TL 	 10 adet satıldı
Product_8 	 C3  		30 TL 	 502 adet satıldı
Product_9 	 C3  		624 TL 	 257 adet satıldı
Product_10 	 C5  		131 TL 	 320 adet satıldı
Product_11 	 C4  		476 TL 	 5 adet satıldı
Product_12 	 C3  		224 TL 	 955 adet satıldı
Product_13 	 C5  		340 TL 	 470 adet satıldı
Product_14 	 C2  		468 TL 	 67 adet satıldı
Product_15 	 C4  		97 TL 	 61 adet satıldı
Product_16 	 C2  		382 TL 	 53 adet satıldı
Product_17 	 C4  		109 TL 	 125 adet satıldı
Product_18 	 C5  		881 TL 	 190 adet satıldı
Product_19 	 C1  		673 TL 	 176 adet satıldı
Product_20 	 C4  		140 TL 	 106 adet satıldı


## Temel İstatistikler

A. Temel İstatistikler

* Ortalama müşteri harcaması nedir? (mean)
* En yüksek ve en düşük harcama miktarları nedir? (max, min)
* Kaç müşteri indirim kullanmıştır? (sum)

In [ ]:
# Temel istatistikler
mean_spending = np.mean(total_spend)      # Ortalama harcama
max_spending = np.max(total_spend)        # En yüksek harcama
min_spending = np.min(total_spend)        # En düşük harcama
discount_users = np.sum(has_discount)     # İndirim kullanan müşteri sayısı

# Sonuçları yazdıralım
print(f"Ortalama müşteri harcaması: {mean_spending:.2f} TL")
print(f"En yüksek harcama: {max_spending} TL")
print(f"En düşük harcama: {min_spending} TL")
print(f"İndirim kullanan müşteri sayısı: {discount_users}")

Ortalama müşteri harcaması: 2578.25 TL
En yüksek harcama: 4998 TL
En düşük harcama: 53 TL
İndirim kullanan müşteri sayısı: 288


##  B. Satış Analizi

1. İndirim kullanan müşterilerin ortalama harcaması ile indirim kullanmayanların ortalama harcaması arasındaki farkı bulun.
2. Toplam satış gelirini hesaplayın.
3. En çok satılan ürünleri ve kategorileri belirleyin.
4. Satın alınan ürün sayısı ile harcama arasında korelasyon var mı? (np.corrcoef() kullanarak)

### B.1

In [ ]:
avg_spending_discounted = np.mean(total_spend[has_discount == 1])
avg_spending_non_discounted = np.mean(total_spend[has_discount == 0])
spending_diff = avg_spending_non_discounted - avg_spending_discounted

print("İndirim kullanan müşterilerin ortalama harcaması", avg_spending_discounted)
print("İndirim kullanmayan müşterilerin ortalama harcaması", avg_spending_non_discounted)
print("İndirim kullanmayanların kullananlardan farkı", spending_diff)

İndirim kullanan müşterilerin ortalama harcaması 2588.6180555555557
İndirim kullanmayan müşterilerin ortalama harcaması 2574.0561797752807
İndirim kullanmayanların kullananlardan farkı -14.561875780274931


### B.2

In [ ]:
total_revenue = np.sum(total_spend)
print("Toplam satış geliri", total_revenue, "TL")

Toplam satış geliri 2578250 TL


### B.3

In [ ]:
product_sales

array([111, 813, 350, 240,  43,  44,  10, 502, 257, 320,   5, 955, 470,
        67,  61,  53, 125, 190, 176, 106])

In [ ]:
# 3. En çok satılan ürün ve kategori
most_sold_product_index = np.argmax(product_sales)   #product_sales dizisindeki en büyük değerin indeksini bulur
most_sold_product = product_names[most_sold_product_index]
most_sold_product_count = product_sales[most_sold_product_index]

most_sold_category = max(category_sales, key=category_sales.get)
most_sold_category_count = category_sales[most_sold_category]

print(f"En çok satılan ürün: {most_sold_product} ({most_sold_product_count} adet)")
print(f"En çok satılan kategori: {most_sold_category} ({most_sold_category_count} adet)")

En çok satılan ürün: Product_12 (955 adet)
En çok satılan kategori: C5 (1007 adet)


### B.4

In [ ]:
product_sales

array([111, 813, 350, 240,  43,  44,  10, 502, 257, 320,   5, 955, 470,
        67,  61,  53, 125, 190, 176, 106])

In [ ]:
#4. Satın alınan ürün sayısı ile harcama arasındaki korelasyon

correlation = np.corrcoef(num_products, total_spend)
print("Satın alınan ürün sayısı ile harcama arasındaki korelasyon:\n", correlation)

Satın alınan ürün sayısı ile harcama arasındaki korelasyon:
 [[1.         0.04342137]
 [0.04342137 1.        ]]


## C. Sezonluk Karşılaştırma (Geçen Yılla Kıyaslama)

Geçen yılın verilerini rastgele oluşturarak bu yıl ile karşılaştırın. Satışlarda ortalama % kaç artış veya azalma olmuş?
Geçen yıl en çok satılan ürün ile bu yıl en çok satılan ürünü karşılaştırın.

### Geçen Yıl için VeriSeti

In [ ]:
# Yeni veri setini üretmek için seed değerini 50 olarak ayarlayarak geçen yılın verilerini oluşturuyoruz

np.random.seed(50)
n_last_year = n  # Geçen yıl da aynı sayıda müşteri olsun

# Geçen yılın müşteri harcama tutarları
total_spend_last_year = np.random.randint(50, 5000, n_last_year)

# Geçen yılın satın alınan ürün sayıları
num_products_last_year = np.random.randint(1, 10, n_last_year)

# Geçen yılın indirim kullanımı durumu
has_discount_last_year = np.random.choice([0, 1], n_last_year, p=[0.7, 0.3])

# Geçen yılın ürün fiyatları
product_prices_last_year = np.random.randint(10, 1000, 20)

# Geçen yılın ürün isimleri
product_names_last_year = np.array([f"Product_{i+1}" for i in range(len(product_prices_last_year))])

# Geçen yılın toplam satılan ürün sayısı
total_products_last_year = np.sum(num_products_last_year)

# Geçen yılın satış adetlerini rastgele belirle, toplam ürün sayısını koruyarak
product_sales_last_year = np.random.multinomial(total_products_last_year, np.random.dirichlet(np.ones(20)))

# Geçen yılın ürün kategorilerini belirle
product_categories_last_year = np.random.choice(categories, total_products_last_year)

# Geçen yılın kategori bazlı satış adetlerini hesapla
unique_categories_last_year, category_sales_counts_last_year = np.unique(product_categories_last_year, return_counts=True)
category_sales_last_year = dict(zip(unique_categories_last_year, category_sales_counts_last_year))

# Sonuçları gösterelim
print(f"Product Name \t Category Name  Price    Sales Count (Last Year)")
for name, price, category, sales in zip(product_names_last_year, product_prices_last_year, product_categories_last_year[:20], product_sales_last_year):
    print(f"{name} \t {category}  \t\t{price} TL \t {sales} adet satıldı")


Product Name 	 Category Name  Price    Sales Count (Last Year)
Product_1 	 C2  		485 TL 	 357 adet satıldı
Product_2 	 C2  		870 TL 	 131 adet satıldı
Product_3 	 C2  		342 TL 	 99 adet satıldı
Product_4 	 C1  		474 TL 	 862 adet satıldı
Product_5 	 C2  		176 TL 	 18 adet satıldı
Product_6 	 C1  		480 TL 	 77 adet satıldı
Product_7 	 C4  		405 TL 	 563 adet satıldı
Product_8 	 C1  		536 TL 	 351 adet satıldı
Product_9 	 C1  		819 TL 	 130 adet satıldı
Product_10 	 C4  		896 TL 	 134 adet satıldı
Product_11 	 C4  		200 TL 	 59 adet satıldı
Product_12 	 C4  		657 TL 	 30 adet satıldı
Product_13 	 C2  		817 TL 	 724 adet satıldı
Product_14 	 C2  		44 TL 	 87 adet satıldı
Product_15 	 C1  		392 TL 	 268 adet satıldı
Product_16 	 C2  		487 TL 	 198 adet satıldı
Product_17 	 C2  		646 TL 	 354 adet satıldı
Product_18 	 C2  		646 TL 	 216 adet satıldı
Product_19 	 C5  		322 TL 	 236 adet satıldı
Product_20 	 C2  		440 TL 	 133 adet satıldı


### Geçen yıl ve bu yılı karşılaştırma

In [ ]:
# 1. Toplam satış adetlerinde yıllık değişim yüzdesi
total_sales_current_year = np.sum(product_sales)
total_sales_last_year = np.sum(product_sales_last_year)
sales_change_percentage = ((total_sales_current_year - total_sales_last_year) / total_sales_last_year) * 100

print("Toplam satış adedi (bu yıl)", total_sales_current_year)
print("Toplam satış adedi (geçen yıl)", total_sales_last_year)
print("Satış değişim yüzdesi", sales_change_percentage)



Toplam satış adedi (bu yıl) 4898
Toplam satış adedi (geçen yıl) 5027
Satış değişim yüzdesi -2.5661428287248857


In [ ]:
# 2. Geçen yıl ve bu yıl en çok satılan ürünleri karşılaştıralım
most_sold_product_current_year_index = np.argmax(product_sales)
most_sold_product_last_year_index = np.argmax(product_sales_last_year)

most_sold_product_current_year = product_names[most_sold_product_current_year_index]
most_sold_product_last_year = product_names_last_year[most_sold_product_last_year_index]

most_sold_product_current_year_count = product_sales[most_sold_product_current_year_index]
most_sold_product_last_year_count = product_sales_last_year[most_sold_product_last_year_index]

print("Bu yıl en çok satılan ürün", most_sold_product_current_year)
print("Bu yıl en çok satılan ürün adedi", most_sold_product_current_year_count)
print("Geçen yıl en çok satılan ürün", most_sold_product_last_year)
print("Geçen yıl en çok satılan ürün adedi", most_sold_product_last_year_count)


Bu yıl en çok satılan ürün Product_12
Bu yıl en çok satılan ürün adedi 955
Geçen yıl en çok satılan ürün Product_4
Geçen yıl en çok satılan ürün adedi 862


### Stratejik Kararlar
Elde edilen verilere dayanarak e-ticaret şirketine 3 öneri sunun. (Örneğin, indirimlerin satışlara etkisini değerlendirerek)

Bu yıl en çok satılan ürün Product_12 (955 adet) olurken, geçen yılın en çok satan ürünü Product_4 (862 adet) idi. Ancak, toplam satışlarda geçen yıla kıyasla %2.56’lık bir düşüş yaşandı. Bu durum, müşteri taleplerinin değişebileceğini ve belirli ürünlere olan ilgide dalgalanmalar olabileceğini gösteriyor. Önerim, geçmiş yılların satış trendlerine bakılarak talebi yüksek ürünlerin stok seviyelerinin doğru planlanmasıdır.

Satış değişim yüzdesi negatif olduğu için, indirimlerin gerçekten satışları artırıp artırmadığı detaylı incelenmeli. Eğer indirim kullanan müşteriler toplam satış gelirine önemli bir katkı sağlamıyorsa, indirim kampanyalarının stratejisi gözden geçirilmeli. Bunun yerine, sadakat programları veya ücretsiz kargo gibi farklı teşvik yöntemleri denenebilir.

Müşteri harcama alışkanlıklarına göre özel fırsatlar sunulabilir.